src/python/pants/util/collections.py

```diff
@@ -71,3 +74,49 @@ def ensure_str_list(val: str | Iterable[str], *, allow_single_str: bool = False)
     If `allow_single_str` is True, a single `str` will be wrapped into a `List[str]`.
     """
     return ensure_list(val, expected_type=str, allow_single_scalar=allow_single_str)
+
+
+def partition_sequentially(
+    items: Sequence[_T],
+    *,
+    key: Callable[[_T], str],
+    size_min: int,
+    size_max: int | None = None,
+) -> Iterator[list[_T]]:
+    """Stably partitions the given items into batches of at least size_min.
+
+    The "stability" property refers to avoiding adjusting all batches when a single item is added,
+    which could happen if the items were trivially windowed using `itertools.islice` and an
+    item was added near the front of the list.
+
+    Batches will be capped to `size_max`, which defaults `size_min*2`.
+    """
+
+    # To stably partition the arguments into ranges of at least `size_min`, we sort them, and
+    # create a new batch sequentially once we have the minimum number of entries, _and_ we encounter
+    # an item hash prefixed with a threshold of zeros.
+    zero_prefix_threshold = math.log(size_min // 8, 2)
+    size_max = size_min * 2 if size_max is None else size_max
+
+    batch: list[_T] = []
+
+    def emit_batch() -> list[_T]:
+        assert batch
```

Comment:
Not sure we idiomatically use `assert` in non-test code?

Reply:
There are a few other non-test instances of `assert` in `util`... but this seems like a safe case: validating that a private function is called correctly.

Reply:
I'm with @stuhood using asserts in normal code is a good way to assert assumptions, document what might otherwise not be necessarily obvious, and of course to coerce mypy.